run/jenkins.py

```diff
@@ -0,0 +1,282 @@
+    # This can be done by running `wpt install firefox` and then verifying
+    # TODO report OS version when creating TestRun
+    # TODO check out correct WPT revision
+
+    # TODO convert these back into command line args. They're currently
+    # env vars due to an earlier change. They should be converted back.
+
+    # TODO break this file up into smaller, unit tested modules.
+    args = {
+        'prod_run': bool(os.environ.get('PROD_RUN', False)),
+        'prod_wet_run': bool(os.environ.get('PROD_WET_RUN', False)),
+        'SHA': os.environ.get('WPT_SHA'),
+        'sauce_from_metadata': bool(os.environ.get('SAUCE_FROM_METADATA')),
+        'sauce_key': os.environ.get('SAUCE_KEY', ''),
+        'sauce_user': os.environ.get('SAUCE_USER', ''),
+        'jenkins_job_name': os.environ.get('JOB_NAME'),
+        'wpt_path': os.environ.get('WPT_PATH'),
+        'wptd_path': '/wptdashboard',
+        'output_path': '/wptdout',
+        'run_path': os.environ.get('RUN_PATH', ''),
+    }
+
+    # Arg validation
+    assert args['wpt_path'], '`WPT_PATH` env var required.'
+    assert not args['wpt_path'].endswith('/'), '`WPT_PATH` cannot end with /.'
+
+    if args['prod_run'] or args['prod_wet_run']:
+        print 'Fetching upload_secret from GCE Metadata...'
+
+        url = '%s/project/attributes/upload_secret' % METADATA_URL
+        res = requests.get(url, headers={'Metadata-Flavor': 'Google'})
+        args['upload_secret'] = res.text
+        assert len(args['upload_secret']) == 64, (
+            'Metadata `upload_secret` must exist for prod runs.')
+
+    platform_id, platform = get_and_validate_platform(args['wptd_path'])
+
+    PROD_HOST = 'https://wptdashboard.appspot.com'
+    GS_RESULTS_BUCKET = 'wptd'
+    LOCAL_LOG_FILEPATH = '%s/wptd-testrun.log' % args['output_path']
+    LOCAL_REPORT_FILEPATH = "%s/wptd-%s-%s-report.log" % (
+        args['output_path'], args['SHA'], platform_id
+    )
+    SUMMARY_PATH = '%s/%s-summary.json.gz' % (args['SHA'], platform_id)
+    LOCAL_SUMMARY_GZ_FILEPATH = "%s/%s" % (args['output_path'], SUMMARY_PATH)
+    GS_RESULTS_FILEPATH_BASE = "%s/%s/%s" % (
+        args['output_path'], args['SHA'], platform_id
+    )
+    GS_HTTP_RESULTS_URL = 'https://storage.googleapis.com/%s/%s' % (
+        GS_RESULTS_BUCKET, SUMMARY_PATH
+    )
+
+    SUMMARY_FILENAME = '%s-%s-summary.json.gz' % (args['SHA'], platform_id)
+    SUMMARY_HTTP_URL = 'https://storage.googleapis.com/%s/%s' % (
+        GS_RESULTS_BUCKET, SUMMARY_FILENAME
+    )
+
+    if platform.get('sauce'):
+        if args['sauce_from_metadata']:
+            print 'Fetching Sauce creds from GCE metadata...'
+
+            for key in ('sauce_user', 'sauce_key'):
+                url = '%s/project/attributes/%s' % (METADATA_URL, key)
+                res = requests.get(url, headers={'Metadata-Flavor': 'Google'})
+                args[key] = res.text
+                assert args[key], 'Metadata key %s is empty.' % key
```

Comment:
Ditto. Is there an obvious reason this might be the case that we can tell the user?

Reply:
Just to clarify in this case, the user == the maintainer of our Jenkins runs

Reply:
Right. My thinking is that if someone new to the project is asked to help manage the code or the Jenkins runs, then useful error messages can probably assist ramp-up and save time.

Reply:
I'll include this in my documentation of how we use metadata.